content/client-lib-development-guide/features.textile

```diff
@@ -267,7 +268,20 @@ h3(#rest-channel). Channel
-** @(RSL6b)@ If, for example, incompatible encryption details are provided or invalid Base64 is detected in the message payload, an error message will be sent to the logger, but the message will still be delivered with last successful decoding and the @encoding@ field. For example, if a message had a decoding of "utf-8/cipher+aes-128-cbc/base64", and the payload was successfully Base64 decoded but the payload could not be encrypted because the @CipherParam@ details were not configured, the message would be delivered with a binary payload and an @encoding@ with the value "utf-8/cipher+aes-128-cbc"
+** @(RSL6b)@ If, for example, incompatible encryption details are provided or invalid Base64 is detected in the message payload, an error message will be sent to the logger, but the message will still be delivered with last successful decoding and the @encoding@ field. For example, if a message had a decoding of "utf-8/cipher+aes-128-cbc/base64", and the payload was successfully Base64 decoded but the payload could not be decrypted because the @CipherParam@ details were not configured, the message would be delivered with a binary payload and an @encoding@ with the value "utf-8/cipher+aes-128-cbc. Additional steps need to be taken if decoding failed on "vcdiff" encoding, see @(RSL6c)@ 
+** @(RSL6c)@ If decoding fails on "vcdiff" encoding that means message delta application failed. Since applying a delta depends on the previous message being present a failure in the delta application on the client side makes it impossible to apply any subsequent deltas received. That would render the channel unusable so a full message needs to be requested from Ably (the server) to replace the failed delta message. That is done by:
+*** @(RSL6c1)@ Following the procedure described in "RTL4c":#RTL4c under the caveats of "RTL4d":#RTL4d, "RTL4e":#RTL4e and "RTL4f":#RTL4f. The @ProtocolMessage.channelSerial@ of the @ATTACH@ @ProtocolMessage@ sent as per "RTL4c":#RTL4c  should be set to the channel serial of the previous message to the message for which "vcdiff" decoding failed.
+*** @(RSL6c2)@ Any subsequent "vcdiff" decoding failures as per "RSL6c":#RSL6c while the recovery procedure from a previous failure as per "RSL6c1":#RSL6c1 is still in progress, i.e. the channel is in @ATTACHING@ state, should not trigger a new recovery procedure as per "RSL6c1":#RSL6c1.
+** @(RSL6d)@ The data payload of the last message on each channel, i.e. @Message.data@ has to be stored at all times since it might be needed to decode the "vcdiff" encoding. The last message bookkeeping rules are:
```

Comment:
> `The data payload of the last message on each channel, i.e. @Message.data@`

This is still technically ambiguous with respect to protocol messages with multiple messages. Maybe `The data payload from the last @Message@ in the @messages@ array of the most recent @ProtocolMessage@ to arrive on the channel (which will have been the most recent @Message@ to be decoded) should be stored at all times, since ...`

(Relatedly: do you think if it might be worth adding a new child of `RSL6` to specify that the messages array must be decoded serially, first-to-last? I can't see anything that actually requires that at the moment, so a client library could theoretically currently be in compliance by decoding them last-to-first, or in parallel)

Reply:
Added an extra point specifying message decoding order.